File: baekjoon/java/src/gold5/Q15681_TreeAndQuery.java
package gold5;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

//https://www.acmicpc.net/problem/15681
public class Q15681_TreeAndQuery {
    private static List<List<Integer>> graph;
    private static int[] numberOfPoint;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] initial = br.readLine().split(" ");
        int n = Integer.parseInt(initial[0]);
        int r = Integer.parseInt(initial[1]);
        int q = Integer.parseInt(initial[2]);

        graph = new ArrayList<>(n+1);
        for (int i = 0; i < n+1; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < n-1; i++) {
            String[] s = br.readLine().split(" ");
            int x = Integer.parseInt(s[0]);
            int y = Integer.parseInt(s[1]);

            graph.get(x).add(y);
            graph.get(y).add(x);
        }

        numberOfPoint = new int[n+1];
        countPoint(r);

        for (int i = 0; i < q; i++) {
            System.out.println(numberOfPoint[Integer.parseInt(br.readLine())]);
        }
    }

    private static void countPoint(int root) {
        numberOfPoint[root] = 1; //부모를 일단 더이상 탐색하지 않게 0이 아니게 만들어줘야함
        for (Integer connectedNode : graph.get(root)) {
            if(numberOfPoint[connectedNode] == 0) {
                countPoint(connectedNode);
                numberOfPoint[root] += numberOfPoint[connectedNode];
            }
        }
    }
}

File: baekjoon/python/gold5/q15681_tree_and_query.py
import sys
from collections import deque
input = sys.stdin.readline
sys.setrecursionlimit(10**9)

## https://www.acmicpc.net/problem/15681
def solution():
    n, r, query = map(int, input().split())
    graph = [[] for _ in range(n+1)]
    for _ in range(n-1):
        x, y = map(int, input().split())
        graph[x].append(y)
        graph[y].append(x)
    
    parent = [0] * (n+1)
    parent[r] = r

    q = deque()
    q.append(r)

    while q :
        now = q.popleft()
        for child in graph[now] :
            if child != parent[now] :
                parent[child] = now
                q.append(child)
    
    subtree = [0] * (n+1)
    subtree[r] = n

    def count_child(root):
        cnt = 1
        if subtree[root] != 0 :
            return subtree[root]
        for child in graph[root]:
            if child != parent[root]:
                cnt += count_child(child)
        subtree[root] = cnt        
        return cnt

    for _ in range(query):
        x = int(input())
        print(count_child(x))

solution()

## https://jainn.tistory.com/m/74
def solution_other() :
    def countPoint(x):
        count[x]=1
        for i in tree[x]:
            if not count[i]:
                countPoint(i)
                count[x] += count[i]

    n, r, q = map(int, input().split())
    tree = [[] for _ in range(n+1)]
    count = [0]*(n+1)

    for i in range(n-1):
        a, b = map(int, input().split())
        tree[a].append(b)
        tree[b].append(a)

    countPoint(r)

    for i in range(q):
        u = int(input())
        print(count[u])

solution_other()
